google-cloud-storage/src/main/java/com/google/cloud/storage/OtelStorageDecorator.java

```diff
@@ -0,0 +1,1719 @@
+    } catch (Throwable t) {
+      sessionSpan.recordException(t);
+      sessionSpan.setStatus(StatusCode.ERROR, t.getClass().getSimpleName());
+      throw t;
+    } finally {
+      sessionSpan.end();
+    }
+  }
+
+  @Override
+  public StorageOptions getOptions() {
+    return delegate.getOptions();
+  }
+
+  static Storage decorate(Storage delegate, OpenTelemetry otel, Transport transport) {
+    requireNonNull(delegate, "delegate must be non null");
+    requireNonNull(otel, "otel must be non null");
+    if (otel == OpenTelemetry.noop()) {
+      return delegate;
+    }
+    Attributes baseAttributes =
+        Attributes.builder()
+            .put("gcp.client.service", "Storage")
+            .put("gcp.client.version", StorageOptions.getDefaultInstance().getLibraryVersion())
+            .put("gcp.client.repo", "googleapis/java-storage")
+            .put("gcp.client.artifact", "com.google.cloud:google-cloud-storage")
+            .put("rpc.system", transport.toString().toLowerCase(Locale.ROOT))
+            .put("service.name", "storage.googleapis.com")
+            .build();
+    return new OtelStorageDecorator(delegate, otel, baseAttributes);
+  }
+
+  private static @NonNull String fmtBucket(String bucket) {
+    return String.format("gs://%s/", bucket);
+  }
+
+  private static final class TracerDecorator implements Tracer {
+    @Nullable private final Context parentContextOverride;
+    private final Tracer delegate;
+    private final Attributes baseAttributes;
+    private final String spanNamePrefix;
+
+    private TracerDecorator(
+        @Nullable Context parentContextOverride,
+        Tracer delegate,
+        Attributes baseAttributes,
+        String spanNamePrefix) {
+      this.parentContextOverride = parentContextOverride;
+      this.delegate = delegate;
+      this.baseAttributes = baseAttributes;
+      this.spanNamePrefix = spanNamePrefix;
+    }
+
+    private static TracerDecorator decorate(
+        @Nullable Context parentContextOverride,
+        OpenTelemetry otel,
+        Attributes baseAttributes,
+        String spanNamePrefix) {
+      requireNonNull(otel, "otel must be non null");
+      requireNonNull(baseAttributes, "baseAttributes must be non null");
+      requireNonNull(spanNamePrefix, "spanNamePrefix must be non null");
+      Tracer tracer =
+          otel.getTracer(OTEL_SCOPE_NAME, StorageOptions.getDefaultInstance().getLibraryVersion());
+      return new TracerDecorator(parentContextOverride, tracer, baseAttributes, spanNamePrefix);
+    }
+
+    @Override
+    public SpanBuilder spanBuilder(String spanName) {
+      SpanBuilder spanBuilder =
+          delegate.spanBuilder(spanNamePrefix + spanName).setAllAttributes(baseAttributes);
+      if (parentContextOverride != null) {
+        spanBuilder.setParent(parentContextOverride);
+      }
+      return spanBuilder;
+    }
+  }
+
+  private static class OtelDecoratedReadChannel implements ReadChannel {
```

Comment:
ReadChannel, WriteChannel, CopyWriter and BlobWriteSession have sub-decorators which carry their originating spans along for their lifetimes.